### PR TITLE
ANSI-714 Multiple Kafka Connect workers/clusters on a single node

### PR DESCRIPTION
**Pull request by nerdynick**
# Description

Adds support for installing multiple Kafka Connect Workers on a single machine/node. With those works also being of separate clusters. 

This also contains bug fixes for Debian when doing Tarball installs.

Replaced PR #332 with a Rebased version

Depends on PR #563

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Single platform install on Ubuntu EC2 VMs

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] Any variable changes have been validated to be backwards compatible

### CODE REVIEW

File: roles/confluent.kafka_connect/tasks/main.yml [hunk 1]
```diff
@@ -57,6 +57,17 @@
     force: true
   when: installation_method == "archive"
 
+# Non-Archive File deployments need to create SystemD service units when the user changes the service name
+- name: Copy Kafka Connect Service from default install to system
+  copy:
+    src: "/lib/systemd/system/{{kafka_connect_default_service_name}}.service"
```
Comment (domenicbove) [inline roles/confluent.kafka_connect/tasks/main.yml:63]: this is hardcoded for ubuntu it looks like

Comment (nerdynick) [reply roles/confluent.kafka_connect/tasks/main.yml]: Fixed

File: roles/confluent.common/tasks/ubuntu.yml [hunk 2]
```diff
@@ -40,6 +40,7 @@
   when:
     - repository_configuration == 'confluent'
     - install_java|bool
+  throttle: 1
```
Comment (domenicbove) [inline roles/confluent.common/tasks/ubuntu.yml:43]: what does this do exactly? 
https://docs.ansible.com/ansible/latest/user_guide/playbooks_strategies.html#restricting-execution-with-throttle

You'll notice in the redhat.yml retries on some tasks that fail.

File: roles/confluent.variables/defaults/main.yml [hunk 3]
```diff
@@ -556,6 +561,12 @@ kafka_rest_monitoring_interceptors_enabled: "{{ monitoring_interceptors_enabled
 
 # Kafka Connect Variables
 
+### Service Name to define/use for Kafka Connect System.d.
+kafka_connect_service_name: "{{kafka_connect_default_service_name}}"
+
+### Config/Properties Filename to use when settingup and configuring Kafka Connect
```
Comment (domenicbove) [inline roles/confluent.variables/defaults/main.yml:567]: "settingup"

File: roles/confluent.kafka_connect/defaults/main.yml [hunk 4]
```diff
@@ -21,7 +21,7 @@ kafka_connect_service_overrides:
   LimitNOFILE: 100000
   User: "{{ kafka_connect_user if kafka_connect_user != kafka_connect_default_user else '' }}"
   Group: "{{ kafka_connect_group if kafka_connect_group != kafka_connect_default_group else '' }}"
-  ExecStart: "{% if installation_method == 'archive' %}{{ kafka_connect.server_start_file }} {{ kafka_connect.config_file }}{% endif %}"
+  ExecStart: "{% if installation_method == 'archive' or kafka_connect_config_filename != kafka_connect_default_config_filename %}{{ kafka_connect.server_start_file }} {{ kafka_connect.config_file }}{% endif %}"
```
Comment (domenicbove) [inline roles/confluent.kafka_connect/defaults/main.yml:24]: nice piggy back on top of the archive logic, this is slick

File: roles/confluent.variables/vars/main.yml [hunk 5]
```diff
@@ -1,6 +1,8 @@
 ---
 confluent_ansible_branch: 6.1.0-post
 
+systemd_base_dir: "{{'/lib/systemd/system' if ansible_distribution == 'Ubuntu' or ansible_distribution == 'Debian' else '/usr/lib/systemd/system'}}"
```
Comment (domenicbove) [inline roles/confluent.variables/vars/main.yml:4]: i think theres an "ansible_family" var that might be equal to Debian that covers both ubuntu and debian

**Review by domenicbove — COMMENTED** [review]
Really good work here! Cant believe it is so simple to make this work. We need a sample inventory file too please

File: roles/confluent.common/tasks/ubuntu.yml (hunk 2 above)
Comment (nerdynick) [reply roles/confluent.common/tasks/ubuntu.yml:43]: It was my solution to a contention on the APT lock when running a deployment on a single node for 2 connect workers. Will switch to the retry to be consistent. Should do the same thing.

File: roles/confluent.variables/vars/main.yml (hunk 5 above)
Comment (nerdynick) [reply roles/confluent.variables/vars/main.yml]: Forgot about that var. Will change out.

File: roles/confluent.common/tasks/ubuntu.yml (hunk 2 above)
Comment (domenicbove) [reply roles/confluent.common/tasks/ubuntu.yml]: omg that makes sense now! makes me think we need to add throttle more places then. There are many yum/apt installs right?

File: sample_inventories/multi_connect_workers_on_single_node.yml [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    
+control_center:
+  hosts:
+    ccc01.confluent.io:
+       
+schema_registry:
+  hosts:
+    sr01.confluent.io:
+                  
+kafka_connect:
```
Comment (domenicbove) [inline sample_inventories/multi_connect_workers_on_single_node.yml:23]: Looks great!!

Comment (domenicbove) [reply sample_inventories/multi_connect_workers_on_single_node.yml:23]: this sample will be super helpful

File: sample_inventories/multi_connect_workers_on_single_node.yml [hunk 7]
```diff
@@ -0,0 +1,49 @@
+all:
```
Comment (domenicbove) [inline sample_inventories/multi_connect_workers_on_single_node.yml:1]:
```suggestion
---
all:
```

to fix yamllint errors in jenkins job

Comment (domenicbove) [reply sample_inventories/multi_connect_workers_on_single_node.yml:1]: actually can you put a blurp at the top in the same format as the others

File: roles/confluent.common/tasks/ubuntu.yml [hunk 8]
```diff
@@ -37,6 +37,9 @@
 - name: Add open JDK repo
   apt_repository:
     repo: "{{ubuntu_java_repository}}"
+  retries: 10
```
Comment (domenicbove) [inline roles/confluent.common/tasks/ubuntu.yml:40]: should this be throttle?
also add a comment on why you are doing throttle

File: roles/confluent.common/tasks/ubuntu.yml [hunk 9]
```diff
@@ -37,6 +37,7 @@
 - name: Add open JDK repo
   apt_repository:
     repo: "{{ubuntu_java_repository}}"
+  throttle: 1 #Throttle was introduced to help mitigate the APT Lock when deploying multiple components to the same node.
```
Comment (domenicbove) [inline roles/confluent.common/tasks/ubuntu.yml:40]:
```suggestion
  # To help mitigate the APT Lock when deploying multiple components to the same node, set throttle to 1
  throttle: 1 
```

Comment (nerdynick) [reply roles/confluent.common/tasks/ubuntu.yml]: Done

**Pull request merged**